src/state/preferences.svelte.ts

```diff
@@ -0,0 +1,49 @@
+import { colors } from '🍎/configs/theme/colors.config.ts';
+import type { WallpaperID } from '🍎/configs/wallpapers/wallpaper.config.ts';
+import { persisted } from './persisted.svelte.ts';
+
+export type WallpaperSettings = {
+  id: WallpaperID;
+  image: string;
+  canControlTheme: boolean;
+};
+
+export type Theme = {
+  scheme: 'light' | 'dark';
+  primaryColor: keyof typeof colors;
+};
+
+export const preferences = persisted('macos:preferences', {
```

Comment:
I feel like trying to abstract `persisted` here isn't doing you any favors because you have to do things like `preferences.value.theme.scheme` instead of `preferences.theme.scheme`. Perhaps you can just create the `$effect` to keep local storage updated directly in this file instead? I think then you don't need the `.value` since you wouldn't be crossing boundaries to access the persisted value

Reply:
I kept it like that cuz Joshnuss is working on svelte-persisted-rune, and once that's out I'm sure he'll figure out a way to do .value on primitives and flattening on objects, so I'll upgrade to that then. Not too worried about the syntax atm

Reply:
I think `preferences.value` is fine — better than a) not having something encapsulated and reusable or b) inconsistency. Unless you were to enforce persisted values being non-primitive

Reply:
I don't know if there's a way around `preferences.value` if you want to have it in a separate file even if you restrict it to being an object? Maybe you could use proxies or something?

You can have a decent amount of reuse if you write it as a hook that the consumer needs to install in an `$effect`. That doesn't buy you much here because this example is so simple, but is potentially something you could do in a more complicated example

Reply:
I figured it out

```ts
import { auto_destroy_effect_root } from './auto-destroy-effect-root.svelte.ts';

type Primitive = string | null | symbol | boolean | number | undefined | bigint;

const is_primitive = (val: any): val is Primitive => {
  return val !== Object(val);
};

export function persisted<T>(key: string, initial: T) {
  const existing = localStorage.getItem(key);

  const primitive = is_primitive(initial);
  const parsed_value = existing ? JSON.parse(existing) : initial;

  let state = $state<T extends Primitive ? { value: T } : T>(
    primitive ? { value: parsed_value } : parsed_value,
  );

  auto_destroy_effect_root(() => {
    $effect(() => {
      // @ts-ignore
      localStorage.setItem(key, JSON.stringify(primitive ? state.value : state));
    });
  });

  return state;
}
```

Reply:
Check the commit just after this comment box

Reply:
> Maybe you could use proxies or something?

No need when svelte does it for ya 😜

Reply:
Nice. That looks much better!

Reply:
Personally don't love the inconsistency but horses for courses!

Tangential to this convo, but: I'm personally of the view that a localStorage abstraction _must_ listen for `storage` events and synchronise between tabs, otherwise one will clobber the other silently, leading to all sorts of fucked up bugs

Reply:
Yet to add that, will do so. I'm also planning on upstreaming the changes to svelte-perisisted-store.